pkg/v1/sdk/features/featuregate/featuregate.go

```diff
@@ -83,23 +83,23 @@ func ComputeFeatureStates(featureGateSpec configv1alpha1.FeatureGateSpec, featur
 	return activated, deactivated, unavailable
 }
 
-// FeatureActivatedInNamespace returns true only if all of the features specified are activated in the namespace.
+// FeatureActivatedInNamespace returns true only if the feature is activated in the namespace.
 func FeatureActivatedInNamespace(ctx context.Context, c client.Client, namespace, feature string) (bool, error) {
-	selector := metav1.LabelSelector{
-		MatchExpressions: []metav1.LabelSelectorRequirement{
-			{Key: "kubernetes.io/metadata.name", Operator: metav1.LabelSelectorOpIn, Values: []string{namespace}},
-		},
-	}
-	return FeaturesActivatedInNamespacesMatchingSelector(ctx, c, selector, []string{feature})
+	return featuresActivated(ctx, c, []string{namespace}, []string{feature})
 }
 
 // FeaturesActivatedInNamespacesMatchingSelector returns true only if all the features specified are activated in every namespace matched by the selector.
+// Note the client passed in must have appropriate RBAC to list namespaces in order to apply the selector.
 func FeaturesActivatedInNamespacesMatchingSelector(ctx context.Context, c client.Client, namespaceSelector metav1.LabelSelector, features []string) (bool, error) {
 	namespaces, err := featureutil.NamespacesMatchingSelector(ctx, c, &namespaceSelector)
 	if err != nil {
 		return false, err
 	}
+	return featuresActivated(ctx, c, namespaces, features)
+}
 
+// featuresActivated returns true only if all the features specified are activated in each input namespace.
+func featuresActivated(ctx context.Context, c client.Client, namespaces, features []string) (bool, error) {
```

Comment:
Is it a valid scenario if two different feature gates have at least one namespace in common? If yes, this function may not work as expected. Please refer the following test:

The test should pass; but it fails.
```
func TestFeatureActivatedInNamespaceUtil(t *testing.T) {
	scheme, err := configv1alpha1.SchemeBuilder.Build()
	if err != nil {
		t.Fatal(err)
	}
	if err := k8sscheme.AddToScheme(scheme); err != nil {
		t.Fatal(err)
	}

	newNamespace := func(name string) *corev1.Namespace {
		return &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: name, Labels: map[string]string{"kubernetes.io/metadata.name": name}}}
	}

	newFeatureGate := func(name string, namespaces, activated, deactivated, unavailable []string) *configv1alpha1.FeatureGate {
		return &configv1alpha1.FeatureGate{
			ObjectMeta: metav1.ObjectMeta{
				Name: name,
			},
			Status: configv1alpha1.FeatureGateStatus{
				Namespaces:          namespaces,
				ActivatedFeatures:   activated,
				DeactivatedFeatures: deactivated,
				UnavailableFeatures: unavailable,
			}}
	}

	testCases := []struct {
		description     string
		existingObjects []runtime.Object
		namespace       []string
		feature         []string
		want            bool
		err             string
	}{
		{
			description: "Namespace exists and is gated and feature is activated",
			existingObjects: []runtime.Object{
				newNamespace("tkg-system"),
				newFeatureGate("tkg-one", []string{"tkg-system"}, []string{"one"}, []string{"three", "four"}, []string{"five", "six"}),
				newFeatureGate("tkg-two", []string{"tkg-system"}, []string{"two"}, []string{"three", "four"}, []string{"five", "six"}),
			},
			namespace: []string{"tkg-system"},
			feature:   []string{"one", "two"},
			want:      true,
			err:       "",
		},
	}

	for _, tc := range testCases {
		t.Run(tc.description, func(t *testing.T) {
			fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithRuntimeObjects(tc.existingObjects...).Build()
			got, err := featuresActivated(context.Background(), fakeClient, tc.namespace, tc.feature)
			if err != nil {
				if tc.err == "" {
					t.Errorf("no error string specified, but got error: %v", err)
				}
				if !strings.Contains(err.Error(), tc.err) {
					t.Errorf("error string=%q doesn't match partial=%q", tc.err, err)
				}
			} else if tc.err != "" {
				t.Errorf("error string=%q specified but error not found", tc.err)
			}
			if got != tc.want {
				t.Errorf("feature activation: got %t, want %t", got, tc.want)
			}
		})
	}
}
```

Reply:
> Is it a valid scenario if two different feature gates have at least one namespace in common?

Great question! This is not a valid scenario. We validate this in the webhook: https://github.com/vmware-tanzu/tanzu-framework/blob/73f58999601aec1660877fd86f470ecba3ede267/apis/config/v1alpha1/featuregate_webhook.go#L118-L119